### PR TITLE
Ensure StorageDead is created even if variable initialization fails

### PR DESCRIPTION
**Pull request by cramertj**
Rebase and slight cleanup of https://github.com/rust-lang/rust/pull/51109
Fixes https://github.com/rust-lang/rust/issues/49232

r? @eddyb

### CODE REVIEW

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! This PR needs your review.

File: src/librustc_mir/build/scope.rs [hunk 1]
```diff
@@ -622,25 +622,47 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
-        } else {
-            // Only temps and vars need their storage dead.
-            match *place {
-                Place::Local(index) if index.index() > self.arg_count => DropKind::Storage,
-                _ => return
+        match drop_kind {
+            DropKind::Value { .. } => if !needs_drop { return },
+            DropKind::Storage => {
+                match *place {
+                    Place::Local(index) if index.index() > self.arg_count => (),
```
[the remark this answers is not in the document]
Comment (cramertj) [reply src/librustc_mir/build/scope.rs]: Done.

**Comment by cramertj** [comment]
@eddyb A bunch of other mir-opt tests fail as a result of this PR due to it introducing duplicate `StorageDead` statements. Is that something I should attempt to fix, or just blindly copy the current output into the expected?

File: src/librustc_mir/build/scope.rs [hunk 2]
```diff
@@ -622,25 +622,54 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+        match drop_kind {
+            DropKind::Value { .. } => if !needs_drop { return },
+            DropKind::Storage => {
+                match *place {
+                    Place::Local(index) => if index.index() <= self.arg_count {
+                        panic!("`schedule_drop` called with index {} and arg_count {}",
+                               index.index(),
+                               self.arg_count)
+                    },
+                    _ => panic!("`schedule_drop` called with non-`Local` place {:?}", place),
```
Comment (eddyb) [inline src/librustc_mir/build/scope.rs]: If you're not using `assert_eq` etc, you can use `bug!` instead of `panic!`.

Comment (cramertj) [reply src/librustc_mir/build/scope.rs]: Switched to `span_bug`.

File: src/librustc_mir/build/scope.rs [hunk 3]
```diff
@@ -895,24 +928,24 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
...
-        // Drop the storage for both value and storage drops.
-        // Only temps and vars need their storage dead.
-        match drop_data.location {
-            Place::Local(index) if index.index() > arg_count => {
-                cfg.push(block, Statement {
-                    source_info,
-                    kind: StatementKind::StorageDead(index)
-                });
+                // Drop the storage for both value and storage drops.
+                // Only temps and vars need their storage dead.
```
Comment (eddyb) [inline src/librustc_mir/build/scope.rs]: You have a check above, I don't think you need another here. You can keep an assert though, and have an `unreachable!()` in the `_ => {}` arm.

File: src/test/mir-opt/issue-49232.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z identify_regions -Z emit-end-regions
```
Comment (eddyb) [inline src/test/mir-opt/issue-49232.rs]: Do we need these two lines? (cc @nikomatsakis @arielb1) I expect the MIR to be shorter and easier to read without them.

**Comment by eddyb** [comment]
r=me with nits fixed

File: src/test/mir-opt/issue-49232.rs (hunk 4 above)
Comment (cramertj) [reply src/test/mir-opt/issue-49232.rs]: Done-- made it a bit shorter.

File: src/librustc_mir/build/scope.rs (hunk 3 above)
Comment (cramertj) [reply src/librustc_mir/build/scope.rs]: Done.

File: src/test/mir-opt/issue-49232.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                false => 4,
+                true => break,
+            }
+        };
+        drop(&beacon);
+    }
+}
+
+// END RUST SOURCE
+// START rustc.main.mir_map.0.mir
```
Comment (eddyb) [inline src/test/mir-opt/issue-49232.rs]: Hmm, what happens if you use the MIR after the first simplifycfg? There should be `mir-opt` examples around that to that. I'm hoping the distinction is not only visible, but much clearer. Right now there are a bunch of pointless blocks being created.

**Pull request merged**